File: Data Structures & Algorithms/combination-target-sum-ii/submission-5.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []
        candidates.sort()

        def f(idx, curr_sum, curr, target):
            if curr_sum == target:
                res.append(curr[:])
                return
            
            if curr_sum > target or idx == len(candidates):
                return

            for i in range(idx + 1, len(candidates)):
                if candidates[i] != candidates[idx]:
                    f(i, curr_sum, curr, target)
                    break
            
            if curr_sum + candidates[idx] <= target:
                curr_sum += candidates[idx]
                curr.append(candidates[idx])
                f(idx + 1, curr_sum, curr, target)

                curr.pop()
                curr_sum -= candidates[idx]




        f(0, 0, [], target)
        return res

File: Data Structures & Algorithms/combination-target-sum/submission-5.java
class Solution {
    public List<List<Integer>> combinationSum(int[] nums, int target) {
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> curr = new ArrayList();
        Arrays.sort(nums);

        f(0, nums, 0, target, res, curr);
        return res;
    }

    private void f(int idx, int[] nums, int sum, int target, List<List<Integer>> res, List<Integer> curr){
        if(idx == nums.length || sum > target){
            return;
        }

        if(sum == target){
            res.add(new ArrayList(curr));
            return;
        }

        curr.add(nums[idx]);
        f(idx, nums, sum + nums[idx], target, res, curr);
        curr.remove(curr.size()-1);
        f(idx +1, nums, sum, target, res, curr);
    }
}
